src/reflect.ts

```diff
@@ -28,9 +32,85 @@ import { AnyAsyncFunction, AnyFunction } from "./util";
  * @param func an in-line ArrowFunction or FunctionExpression. It must be in-line and cannot reference
  *             a variable or a computed function/closure.
  */
-export declare function reflect<F extends AnyFunction | AnyAsyncFunction>(
+export function reflect<F extends AnyFunction | AnyAsyncFunction>(
   func: F
-): FunctionDecl<F> | Err;
+): FunctionLike<F> | Err | undefined {
+  if (func.name.startsWith("bound ")) {
+    // native bound function
+    const targetFunc = (<any>func)[ReflectionSymbols.TargetFunction];
+    if (targetFunc) {
+      return reflect(targetFunc);
+    } else {
+      throw new Error(
+        `bound function not compiled with Functionless, cannot introspect on its form`
+      );
+    }
+  }
+
+  const astCallback = (<any>func)[ReflectionSymbols.AST];
+  if (typeof astCallback === "function") {
+    if (!reflectCache.has(astCallback)) {
+      reflectCache.set(
+        astCallback,
+        parseSExpr(astCallback()) as FunctionLike | Err
+      );
+    }
+    return reflectCache.get(astCallback) as FunctionLike<F> | Err | undefined;
+  }
+  return undefined;
+}
+
+const Global: any = global;
+
+// @ts-ignore
+const reflectCache: WeakMap<Function, FunctionLike | Err> = (Global[
+  Symbol.for("functionless:Reflect")
+] = Global[Symbol.for("functionless:Reflect")] ?? new WeakMap());
 
 // to prevent the closure serializer from trying to import all of functionless.
 export const deploymentOnlyModule = true;
+
+export const ReflectionSymbolNames = {
+  AST: "functionless:AST",
+  BoundThis: "functionless:BoundThis",
+  BoundArgs: "functionless:BoundArgs",
+  TargetFunction: "functionless:TargetFunction",
+} as const;
+
+export const ReflectionSymbols = {
+  AST: Symbol.for(ReflectionSymbolNames.AST),
+  BoundThis: Symbol.for(ReflectionSymbolNames.BoundThis),
+  BoundArgs: Symbol.for(ReflectionSymbolNames.BoundArgs),
+  TargetFunction: Symbol.for(ReflectionSymbolNames.TargetFunction),
+} as const;
+
+export const isFunctionLikeOrErr = anyOf(isFunctionLike, isErr);
+
+export function validateFunctionLike(
+  a: any,
+  functionLocation: string
+): FunctionLike {
+  return validateFunctionlessNode(a, functionLocation, isFunctionLike);
+}
+
+function validateFunctionlessNode<E extends FunctionlessNode>(
+  a: any,
+  functionLocation: string,
+  validate: (e: FunctionlessNode) => e is E
+): E {
+  if (validate(a)) {
+    return a;
+  } else if (isErr(a)) {
+    throw a.error;
+  } else if (typeof a === "function") {
+    if (a.name.startsWith("bound")) {
+    }
```

Comment:
????????

Reply:
Are we missing tests for this code?

Reply:
the code isn't broken, i just decided to move the logic to `reflect` instead of here and forgot to clean up

Reply:
will add a test specifically for bound function